main.tf

```diff
@@ -90,7 +98,46 @@ resource "azurerm_subnet_route_table_association" "hub_routing_external" {
   subnet_id      = each.value.subnet_id
 }
 
-#module "azure_firewall" {
-#  source = "..."
-#  # TODO
-#}
+resource "azurerm_public_ip" "fw_default_ip_configuration_pip" {
+  for_each = local.fw_default_ip_configuration_pip
+
+  allocation_method   = "Static"
+  location            = each.value.location
+  name                = each.value.name
+  resource_group_name = each.value.resource_group_name
+  ip_version          = each.value.ip_version
+  sku                 = each.value.sku
+  sku_tier            = each.value.sku_tier
+  tags                = {}
+  zones               = each.value.zones
+}
+
+locals {
+  vnet_firewall_default_ip_configuration_public_ip_id = {
+    for vnet_name, pip in azurerm_public_ip.fw_default_ip_configuration_pip : vnet_name => pip.id
+  }
+}
+
+resource "azurerm_firewall" "fw" {
+  for_each = local.firewalls
+
+  location            = module.hub_virtual_networks[each.key].vnet_location
+  name                = each.value.name
+  resource_group_name = var.hub_virtual_networks[each.key].resource_group_name
+  sku_name            = each.value.sku_name
+  sku_tier            = each.value.sku_tier
+  tags                = {}
+
```

Comment:
I think we should support these to enable the functionality people will want:

- https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/firewall#private_ip_ranges
- https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/firewall#threat_intel_mode
- https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/firewall#zones
- https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/firewall#dns_servers
- https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/firewall#firewall_policy_id

Reply:
As for `firewall_policy_id`, since the resource group is created by us, I guess that the policy must be created by us too. I'll add this policy parameter as an embedded parameter inside firewall variable. @matt-FFFFFF

Reply:
I've changed my mind... Creating `azurerm_firewall_policy` would lead to the creation of user assigned identity, which requires the creation of resource group too. I think that'll make this module way too much complex. I just recall that we can create resource group outside this module, so I think we'd better leave user assigned identity / firewall policy out of this module. Just a simple firewall, if the user is not satisfied with that, then just create his own resource along with the firewall, firewall policy, or identity. WDYT?

Reply:
Hi @lonegunmanb 

In reality we probably don't want to be responsible for creating the firewall policy in this module.

A customer can link it to an existing one. One a policy is not supplied then we don't make the link.

Therefore we should remove the azurerm_firewall_policy from this module and leave it to be supplied by the caller.

Reply:
So, we create the firewall infrastructure as we need to know the internal IP address for routing purposes.

But, the firewall administration (the policy) is managed externally to this module.